### PR TITLE
Cart show

### LINKED ISSUE #47: User Story 26, User views their cart show page with items in the cart
```
As a visitor or registered user
When I have added items to my cart
And I visit my cart ("/cart")
I see all items I've added to my cart
And I see a link to empty my cart
Each item in my cart shows the following information:
- the name of the item
- a very small thumbnail image of the item
- the merchant I'm buying this item from
- the price of the item
- my desired quantity of the item
- a subtotal (price multiplied by quantity)

I also see a grand total of what everything in my cart will cost
```

### Checklist:

- [ ] base tests are written
- [ ] base code is written to pass tests
- [ ] edge case ("sad path") tests written (if required)
- [ ] edge case code written (if required)
- [ ] code is reviewed
- [ ] branch is merged
- [ ] all tests still pass

child of #49

child of #79

depends on #76

### PR DESCRIPTION
**Pull request by teresa-m-knowles**
This completes user story #26 and closes #47 in waffle

>  As a visitor or registered user
> When I have added items to my cart
> And I visit my cart ("/cart")
> I see all items I've added to my cart
> And I see a link to empty my cart
> Each item in my cart shows the following information:
> - the name of the item
> - a very small thumbnail image of the item
> - the merchant I'm buying this item from
> - the price of the item
> - my desired quantity of the item
> - a subtotal (price multiplied by quantity)
> 
> I also see a grand total of what everything in my cart will cost
> 

I had some issues pulling logic from the view, which is why I'm opening this as a draft PR. I'll comment on places where we need to refactor, since it's a little bit hidden in the Bootstrap. I welcome any ideas as to how to make this one better!

Here's how it ends up looking from a CSS point of view:

screen shot 2019-02-15 at 2 29 11 pm

### CODE REVIEW

File: app/controllers/carts_controller.rb [hunk 1]
```diff
@@ -5,6 +5,11 @@ class CartsController < ApplicationController
 
   def show
     flash[:primary] = "Your cart is empty." if @cart.total_count == 0
+    set_cart
+    @items = @cart.contents.map do |item_id, quantity|
```
Comment (teresa-m-knowles) [inline app/controllers/carts_controller.rb]: This gets me all the items in the cart so that I can iterate through them in the cart show page and pull info, like item.name, etc

File: app/models/cart.rb [hunk 2]
```diff
@@ -14,4 +14,11 @@ def total_count
   def add_item(id)
     @contents[id] = (@contents[id].to_i + 1).to_s
   end
+
+  def total
```
Comment (teresa-m-knowles) [inline app/models/cart.rb:18]: grand total for a cart

File: app/models/item.rb [hunk 3]
```diff
@@ -62,4 +62,8 @@ def average_fulfillment_time
   def ordered?
     order_items.count > 0
   end
+
```
Comment (teresa-m-knowles) [inline app/models/item.rb:65]: If I pass a quantity to it, this gets me the subtotal for that specific item

File: app/views/carts/show.html.erb [hunk 4]
```diff
@@ -0,0 +1,61 @@
+<div class="container">
+
+
+  <% @items.each do |item| %>
+  <% quantity = @cart.contents[item.id.to_s].to_i %>
```
Comment (teresa-m-knowles) [inline app/views/carts/show.html.erb]: This feels wroooong, but I couldn't figure it out any other way. Essentially, I want to grab all of the items from the cart and iterate through them. The problem is that the quantities are not in the @items array, but in the @cart.contents values, so I had to come up with this workaround. 

This is the part where I need the most help. I tried to iterate through the @cart.contents hash instead, but that then left me with items.ids, not Items. Let me know if that makes sense and what would be a good workaround.

File: app/views/carts/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      </div>
+      <div class="col-md-4">
+        <h4 class="product-name"><strong> <%= item.name %> </strong></h4>
+        <h4><small> <%= item.description %> </small></h4>
+        <h4><small> <strong>Merchant: </strong><%= item.user.name %> </small></h4>
+      </div>
+      <div class="col-md-4">
+          <p><strong>Price:</strong>  <%= number_to_currency(item.price)%></p>
+          <p><strong>Quantity:</strong> <%= quantity %></p>
+          <p><strong>Subtotal:</strong> <%= number_to_currency(item.subtotal(quantity)) %></p>
```
Comment (teresa-m-knowles) [inline app/views/carts/show.html.erb:17]: Here's where I use that quantity variable to get the subtotal.

File: app/views/carts/show.html.erb (hunk 4 above)
Comment (allieIsOkay) [reply app/views/carts/show.html.erb]: It might be a bit of an overthink, but if we wanted to change it we could do an each on the cart instead of a map and build a hash out with it. Building this as a model method might also be a possibility. Something like the below might work:
``` ruby
@items = {}
@cart.contents.each do |item, quantity|
   @items[Item.find(item)] = quantity
end
```

This would give us something we could iterate over that has item objects, and the quantities associated with them. I'd love to hear everyone's thoughts.

**Comment by allieIsOkay** [comment]
It looks absolutely great on the page! I'm not entirely sure what the best approach would be for the quantity issue.

File: app/views/carts/show.html.erb (hunk 4 above)
Comment (teresa-m-knowles) [reply app/views/carts/show.html.erb]: Yes! That's a great solution! Let me try that on my branch and push it, unless someone else has another idea.

**Comment by csvlewis** [comment]
The Bootstrap styling looks great! To be honest, it's kind of hard for me to suggest refactors because I haven't worked on any of the cart functionality in the project yet. Maybe we can walk through the code together tomorrow at Turing and think up some ideas?

File: app/controllers/carts_controller.rb [hunk 6]
```diff
@@ -5,6 +5,11 @@ class CartsController < ApplicationController
 
   def show
     flash[:primary] = "Your cart is empty." if @cart.total_count == 0
+    set_cart
+    @items = {}
+    @cart.contents.each do |item, quantity|
```
Comment (teresa-m-knowles) [inline app/controllers/carts_controller.rb:10]: @plapicola this worked great! Thanks again

File: app/views/carts/show.html.erb (hunk 4 above)
Comment (teresa-m-knowles) [reply app/views/carts/show.html.erb]: @plapicola Just implemented that change, it worked great. Changing this from a draft to a pr

**Review by allieIsOkay — APPROVED** [review]
Looks fantastic, great work getting the bootstrap in out of the gate.

**Comment by timnallen** [comment]
Great job! This looks really great! The bootstrap is excellent.
